tests/acceptance/comparative_booster_study/variables_lib.py

```diff
@@ -0,0 +1,140 @@
+import operator
+from functools import reduce
+
+from databuilder.codes import CTV3Code, ICD10Code
+from databuilder.query_language import case, when
+
+from . import schema
+
+
+def create_sequential_variables(
+    dataset, variable_name_template, events, column, num_variables, sort_column=None
+):
+    sort_column = sort_column or column
+    for index in range(num_variables):
+        next_event = events.sort_by(getattr(events, sort_column)).first_for_patient()
+        events = events.take(
+            getattr(events, sort_column) > getattr(next_event, sort_column)
+        )
+        variable_name = variable_name_template.format(n=index + 1)
+        setattr(dataset, variable_name, getattr(next_event, column))
+
+
+def _registrations_overlapping_period(start_date, end_date):
+    regs = schema.practice_registrations
+    return regs.take(
+        regs.start_date.is_on_or_before(start_date)
+        & (regs.end_date.is_after(end_date) | regs.end_date.is_null())
+    )
+
+
+def practice_registration_as_of(date):
+    regs = _registrations_overlapping_period(date, date)
+    return regs.sort_by(regs.start_date, regs.end_date).first_for_patient()
+
+
+def was_registered_with_one_practice_between(start_date, end_date):
+    return _registrations_overlapping_period(start_date, end_date).exists_for_patient()
+
+
+def date_deregistered_from_all_supported_practices():
+    max_dereg_date = schema.practice_registrations.end_date.maximum_for_patient()
+    # In TPP currently active registrations are recorded as having an end date of
+    # 9999-12-31. We convert these, and any other far-future dates, to NULL.
```

Comment:
It would certainly be nice to not have to worry about it here. But, unless I'm missing an obvious trick, I don't think there's an efficient way to do this other than changing what's actually in the database, which isn't a near term option for us in TPP.

I'm pretty sure that wrapping every date column reference in a case expression to convert `9999-12-31` to NULL will prevent any use of the index.

Reply:
It does a bit. Although I'm hoping that with new backend providers we'll be able to get them to reshape the data, or use materialised views or whatever, such that these kinds of issue don't arise. And with some backends, e.g Presto/Trino, everything's a sequential scan anyway so it probably doesn't make much difference.

So I think this is quite a TPP-specific issue for us, and that makes me more relaxed about having this kind of workaround for things that ideally would be in the contract.